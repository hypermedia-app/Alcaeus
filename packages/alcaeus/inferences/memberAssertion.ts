import type { DatasetCore, Quad } from '@rdfjs/types'
import { hydra } from '@tpluscode/rdf-ns-builders'
import { HydraEnvironment } from 'alcaeus-core'

export function * addExplicitStatementsInferredFromMemberAssertion(dataset: DatasetCore, env: HydraEnvironment): Iterable<Quad> {
  for (const collection of env.clownface({ dataset }).has([hydra.manages, hydra.memberAssertion]).toArray()) {
    const assertions = collection.out([hydra.manages, hydra.memberAssertion]).toArray()

    for (const member of collection.out(hydra.member).toArray()) {
      for (const assertion of assertions) {
        let blanks = 0

        let subject = assertion.out(hydra.subject)
        let predicate = assertion.out(hydra.property)
        let object = assertion.out(hydra.object)

        if (subject.terms.length === 0) {
          subject = member
          blanks++
        }

        if (predicate.terms.length === 0) {
          predicate = member
          blanks++
        }

        if (object.terms.length === 0) {
          object = member
          blanks++
        }

        if (blanks === 1) {
          for (const s of subject.terms) {
            for (const p of predicate.terms) {
              for (const o of object.terms) {
                yield env.quad(<any>s, <any>p, <any>o)
              }
            }
          }
        }
      }
    }
  }
}
