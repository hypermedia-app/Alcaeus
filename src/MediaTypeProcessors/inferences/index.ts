export { inferTypesFromPropertyRanges } from './propertyTypes'
export { addExplicitStatementsInferredFromManagesBlock } from './managesBlock'
export { inheritFromSuperclasses } from './subclassOfInherit'
