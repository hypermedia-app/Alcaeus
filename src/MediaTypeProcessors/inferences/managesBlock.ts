import $rdf from 'rdf-ext'
import { Core } from '../../Constants'

function countNotNull (...args: (object | null)[]) {
    return args.reduce((previousValue, currentValue) => {
        if (currentValue) {
            return previousValue + 1
        }

        return previousValue
    }, 0)
}

export function addExplicitStatementsInferredFromManagesBlock (dataset) {
    const managesBlocks = dataset.match(null, $rdf.namedNode(Core.Vocab('manages')), null, null).toArray()
        .map(manages => {
            const collection = manages.subject
            const objectMatches = dataset.match(manages.object, $rdf.namedNode(Core.Vocab('object'), null, null))
            const subjectMatches = dataset.match(manages.object, $rdf.namedNode(Core.Vocab('subject'), null, null))
            const propertyMatches = dataset.match(manages.object, $rdf.namedNode(Core.Vocab('property'), null, null))

            return {
                collection,
                subject: subjectMatches.length === 1 ? subjectMatches.toArray()[0].object : null,
                property: propertyMatches.length === 1 ? propertyMatches.toArray()[0].object : null,
                object: objectMatches.length === 1 ? objectMatches.toArray()[0].object : null,
            }
        })

    managesBlocks.forEach(({ collection, object, property, subject }) => {
        if (countNotNull(object, property, subject) !== 2) {
            return
        }

        const members = dataset.match(collection, $rdf.namedNode(Core.Vocab('member'), null, null))

        members.forEach(memberQuad => {
            const member = memberQuad.object
            dataset.add($rdf.triple(
                subject || member,
                property || member,
                object || member,
            ))
        })
    })
}
