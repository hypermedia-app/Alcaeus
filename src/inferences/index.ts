export { inferTypesFromPropertyRanges } from './propertyTypes.js'
export { addExplicitStatementsInferredFromMemberAssertion } from './memberAssertion.js'
