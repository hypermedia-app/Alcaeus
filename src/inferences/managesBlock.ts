import cf, { Clownface } from 'clownface'
import { DatasetCore } from 'rdf-js'
import { hydra } from '@tpluscode/rdf-ns-builders'

interface ManagesBlockAssertion {
    subject: Clownface
    predicate: Clownface
    object: Clownface
}

export function addExplicitStatementsInferredFromManagesBlock(dataset: DatasetCore) {
    cf({ dataset }).has(hydra.manages).forEach(collection => {
        const managesBlocks = collection.out(hydra.manages)

        collection.out(hydra.member).forEach(member => {
            managesBlocks.forEach(managesBlock => {
                let blanks = 0

                let subject = managesBlock.out(hydra.subject)
                let predicate = managesBlock.out(hydra.property)
                let object = managesBlock.out(hydra.object)

                if (subject.terms.length === 0) {
                    subject = member
                    blanks++
                }

                if (predicate.terms.length === 0) {
                    predicate = member
                    blanks++
                }

                if (object.terms.length === 0) {
                    object = member
                    blanks++
                }

                if (blanks === 1) {
                    subject.addOut(predicate, object)
                }
            })
        })
    })
}
