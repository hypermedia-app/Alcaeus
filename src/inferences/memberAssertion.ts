import type { BaseQuad, DatasetCore } from '@rdfjs/types'
import cf from 'clownface'
import * as RDF from '@rdf-esm/data-model'
import { hydra } from '@tpluscode/rdf-ns-builders'

export function * addExplicitStatementsInferredFromMemberAssertion(dataset: DatasetCore): Iterable<BaseQuad> {
    for (const collection of cf({ dataset }).has([hydra.manages, hydra.memberAssertion]).toArray()) {
        const assertions = collection.out([hydra.manages, hydra.memberAssertion]).toArray()

        for (const member of collection.out(hydra.member).toArray()) {
            for (const assertion of assertions) {
                let blanks = 0

                let subject = assertion.out(hydra.subject)
                let predicate = assertion.out(hydra.property)
                let object = assertion.out(hydra.object)

                if (subject.terms.length === 0) {
                    subject = member
                    blanks++
                }

                if (predicate.terms.length === 0) {
                    predicate = member
                    blanks++
                }

                if (object.terms.length === 0) {
                    object = member
                    blanks++
                }

                if (blanks === 1) {
                    for (const s of subject.terms) {
                        for (const p of predicate.terms) {
                            for (const o of object.terms) {
                                yield RDF.quad<BaseQuad>(s, p, o)
                            }
                        }
                    }
                }
            }
        }
    }
}
