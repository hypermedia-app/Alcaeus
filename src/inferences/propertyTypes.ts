import cf from 'clownface'
import { DatasetCore } from 'rdf-js'
import { hydra, rdf } from '@tpluscode/rdf-ns-builders'

const propertyRangeMappings = [
    [hydra.supportedClass, hydra.Class],
    [hydra.expects, hydra.Class],
    [hydra.returns, hydra.Class],
    [hydra.supportedOperation, hydra.SupportedOperation],
    [hydra.operation, hydra.Operation],
    [hydra.supportedProperty, hydra.SupportedProperty],
    [hydra.statusCodes, hydra.StatusCodeDescription],
    [hydra.property, rdf.Property],
    [hydra.mapping, hydra.IriTemplateMapping],
]

export function inferTypesFromPropertyRanges(dataset: DatasetCore) {
    const node = cf({ dataset })

    propertyRangeMappings.map(([ property, type ]) => {
        node.out(property).addOut(rdf.type, type)
    })
}
