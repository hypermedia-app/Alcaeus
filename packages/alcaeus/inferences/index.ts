import { inferTypesFromPropertyRanges } from './propertyTypes.js'
import { addExplicitStatementsInferredFromMemberAssertion } from './memberAssertion.js'

export default [
  inferTypesFromPropertyRanges,
  addExplicitStatementsInferredFromMemberAssertion,
]
