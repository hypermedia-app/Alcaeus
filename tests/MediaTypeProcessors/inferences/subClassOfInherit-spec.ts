import * as graphs from './subClassOfInherit-spec-graphs'
import { inheritFromSuperclasses } from '../../../src/MediaTypeProcessors/inferences'

describe('subClassOf inference', () => {
    it('reasserts supported operations from superclasses', async () => {
        // given
        const dataset = await graphs.multiLevelSupportedOperations()

        // when
        inheritFromSuperclasses(dataset)

        // then
        expect(dataset.toCanonical()).toMatchSnapshot()
    })

    it('reasserts supported properties from superclasses', async () => {
        // given
        const dataset = await graphs.multiLevelSupportedProperties()

        // when
        inheritFromSuperclasses(dataset)

        // then
        expect(dataset.toCanonical()).toMatchSnapshot()
    })
})
