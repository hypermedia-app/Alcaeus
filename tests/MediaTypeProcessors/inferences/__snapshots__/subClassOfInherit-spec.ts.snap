// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`subClassOf inference reasserts supported operations from superclasses 1`] = `
"<http://example.com/vocab#BaseClass> <http://www.w3.org/1999/02/22-rdf-syntax-ns#type> <http://www.w3.org/ns/hydra/core#Class> .
<http://example.com/vocab#BaseClass> <http://www.w3.org/ns/hydra/core#supportedOperation> <http://example.com/api#HeadRequest> .
<http://example.com/vocab#DraftIssue> <http://www.w3.org/1999/02/22-rdf-syntax-ns#type> <http://www.w3.org/ns/hydra/core#Class> .
<http://example.com/vocab#DraftIssue> <http://www.w3.org/2000/01/rdf-schema#subClassOf> <http://example.com/vocab#Issue> .
<http://example.com/vocab#DraftIssue> <http://www.w3.org/ns/hydra/core#supportedOperation> <http://example.com/api#GetIssue> .
<http://example.com/vocab#DraftIssue> <http://www.w3.org/ns/hydra/core#supportedOperation> <http://example.com/api#HeadRequest> .
<http://example.com/vocab#DraftIssue> <http://www.w3.org/ns/hydra/core#supportedOperation> <http://example.com/api#UpdateDraft> .
<http://example.com/vocab#Issue> <http://www.w3.org/1999/02/22-rdf-syntax-ns#type> <http://www.w3.org/ns/hydra/core#Class> .
<http://example.com/vocab#Issue> <http://www.w3.org/2000/01/rdf-schema#subClassOf> <http://example.com/vocab#BaseClass> .
<http://example.com/vocab#Issue> <http://www.w3.org/ns/hydra/core#supportedOperation> <http://example.com/api#GetIssue> .
<http://example.com/vocab#Issue> <http://www.w3.org/ns/hydra/core#supportedOperation> <http://example.com/api#HeadRequest> .
"
`;

exports[`subClassOf inference reasserts supported properties from superclasses 1`] = `
"<http://example.com/vocab#BaseClass> <http://www.w3.org/1999/02/22-rdf-syntax-ns#type> <http://www.w3.org/ns/hydra/core#Class> .
<http://example.com/vocab#BaseClass> <http://www.w3.org/ns/hydra/core#supportedProperty> _:c14n0 .
<http://example.com/vocab#DraftIssue> <http://www.w3.org/1999/02/22-rdf-syntax-ns#type> <http://www.w3.org/ns/hydra/core#Class> .
<http://example.com/vocab#DraftIssue> <http://www.w3.org/2000/01/rdf-schema#subClassOf> <http://example.com/vocab#Issue> .
<http://example.com/vocab#DraftIssue> <http://www.w3.org/ns/hydra/core#supportedProperty> _:c14n0 .
<http://example.com/vocab#DraftIssue> <http://www.w3.org/ns/hydra/core#supportedProperty> _:c14n1 .
<http://example.com/vocab#DraftIssue> <http://www.w3.org/ns/hydra/core#supportedProperty> _:c14n2 .
<http://example.com/vocab#DraftIssue> <http://www.w3.org/ns/hydra/core#supportedProperty> _:c14n3 .
<http://example.com/vocab#Issue> <http://www.w3.org/1999/02/22-rdf-syntax-ns#type> <http://www.w3.org/ns/hydra/core#Class> .
<http://example.com/vocab#Issue> <http://www.w3.org/2000/01/rdf-schema#subClassOf> <http://example.com/vocab#BaseClass> .
<http://example.com/vocab#Issue> <http://www.w3.org/ns/hydra/core#supportedProperty> _:c14n0 .
<http://example.com/vocab#Issue> <http://www.w3.org/ns/hydra/core#supportedProperty> _:c14n2 .
<http://example.com/vocab#Issue> <http://www.w3.org/ns/hydra/core#supportedProperty> _:c14n3 .
_:c14n0 <http://www.w3.org/ns/hydra/core#property> <http://example.com/vocab#title> .
_:c14n1 <http://www.w3.org/ns/hydra/core#property> <http://example.com/vocab#hasBeenReviewed> .
_:c14n2 <http://www.w3.org/ns/hydra/core#property> <http://example.com/vocab#assignee> .
_:c14n3 <http://www.w3.org/ns/hydra/core#property> <http://example.com/vocab#status> .
"
`;
